v1_21_R2/src/main/java/net/citizensnpcs/nms/v1_21_R2/entity/EntityHumanNPC.java

```diff
@@ -86,6 +91,25 @@ public EntityHumanNPC(MinecraftServer minecraftServer, ServerLevel world, GamePr
         }
     }
 
+    private boolean canCallNPCMoveEvent;
+    @Override
+    public void aiStep() {
+        canCallNPCMoveEvent = true;
+        try {
+            super.aiStep();
+        } finally {
+            canCallNPCMoveEvent = false;
+        }
+    }
+
+    @Override
+    protected void pushEntities() {
+        super.pushEntities();
+        if (canCallNPCMoveEvent) {
+            NMSImpl.callNPCMoveEvent(this);
```

Comment:
Why is called here?

Reply:
we don't have mixin-like things. as the logic you've provided, paper entity move event is called after pushing entities but before drown hurt. I'm just emulating that.

Reply:
We call pushEntities in Citizens as well, is this tested?

Reply:
> We call pushEntities in Citizens as well, is this tested?

That is fine as you don't call aistep. Because of the flag is only updated by aistep method so any other calls to pushEntities will not call move events.

Reply:
After some deep research on tick code we use, aistep is not called so this is invalid.
And because of we took over the control of the whole tick code of NPC so I could call events directly in tick method.
But this is not applicable to other entity types.